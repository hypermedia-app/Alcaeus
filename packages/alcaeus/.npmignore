docs/
gitbook/
src/
*.config.js
.idea/
.github/
tests/
*.json
.travis.yml
.editorconfig
*ignore
coverage/
