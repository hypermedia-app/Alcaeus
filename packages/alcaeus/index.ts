import { RdfineFactory } from '@tpluscode/rdfine'
import { HydraFactory } from '@rdfine/hydra/Factory'
import { AlcaeusHydraFactory } from 'alcaeus-model/Factory.js'
import create from './Factory.js'

export type { ResourceIdentifier, ResourceIndexer, ResourceFactory, RdfResource } from '@tpluscode/rdfine'
export * from 'alcaeus-model'
export type { RuntimeOperation } from 'alcaeus-model/Operation.js'
export type { HydraResponse } from './alcaeus.js'

export default function (init: Parameters<typeof create>[0] = {}) {
  return [RdfineFactory, HydraFactory, AlcaeusHydraFactory, create(init)]
}
