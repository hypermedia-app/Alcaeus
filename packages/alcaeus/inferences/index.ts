import { inferTypesFromPropertyRanges } from './propertyTypes.js'
import { addExplicitStatementsInferredFromMemberAssertion } from './memberAssertion.js'

export default [
  inferTypesFromPropertyRanges,
  addExplicitStatementsInferredFromMemberAssertion,
]
