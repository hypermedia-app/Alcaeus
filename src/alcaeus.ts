import type {DatasetCore, NamedNode} from '@rdfjs/types'
import type {RdfResourceCore} from '@tpluscode/rdfine/RdfResource'
import cf from 'clownface'
import type {GraphPointer} from 'clownface'
import type {Resource, Operation, ApiDocumentation} from '@rdfine/hydra'
import {hydra} from '@tpluscode/rdf-ns-builders'
import fromStream from 'rdf-dataset-ext/fromStream.js'
import {ResourceIdentifier} from '@tpluscode/rdfine'
import $rdf, {HydraEnvironment} from './environment.js'
import FetchUtil, {AllowedRequestInit} from './FetchUtil.js'
import {merge} from './helpers/MergeHeaders.js'
import {getAbsoluteUri} from './helpers/uri.js'
import * as DefaultCacheStrategy from './ResourceCacheStrategy.js'
import type {ResourceCacheStrategy} from './ResourceCacheStrategy.js'
import type {ResourceRepresentation} from './ResourceRepresentation.js'
import type {ResourceStore} from './ResourceStore.js'
import type {ResponseWrapper} from './ResponseWrapper.js'
import type {RootNodeCandidate} from './RootSelectors/index.js'

type InvokedOperation = Pick<Operation, 'method'> & {
  target: Pick<Resource, 'id'>
}

export interface HydraResponse<D extends DatasetCore = DatasetCore, T extends RdfResourceCore<D> = Resource<D>> {
  representation?: ResourceRepresentation<D, T>
  response?: ResponseWrapper
}

function byInProperties(left: GraphPointer, right: GraphPointer) {
  return left.in().terms.length - right.in().terms.length
}

interface Defaults<T> {
  (arg: { uri: string }): T | Promise<T>
}

export interface HydraClient<D extends DatasetCore = DatasetCore> {
  log: (msg: string) => void
  baseUri?: string
  rootSelectors: [string, RootNodeCandidate][]

  loadResource<T extends RdfResourceCore<any> = Resource<D>>(uri: string | NamedNode, headers?: HeadersInit, request?: AllowedRequestInit): Promise<HydraResponse<D, T>>

  loadDocumentation(uri: string | NamedNode, headers?: HeadersInit, request?: AllowedRequestInit): Promise<ApiDocumentation<D> | null>

  invokeOperation<T extends RdfResourceCore<any> = Resource<D>>(operation: InvokedOperation, headers?: HeadersInit, body?: BodyInit, request?: AllowedRequestInit): Promise<HydraResponse<D, T>>

  defaultHeaders: HeadersInit | Defaults<HeadersInit>
  defaultRequestInit: AllowedRequestInit | Defaults<AllowedRequestInit>
  resources: ResourceStore<D>
  apiDocumentations: ResourceRepresentation<D, ApiDocumentation<D>>[]
  cacheStrategy: ResourceCacheStrategy
}

interface AlcaeusInit<D extends DatasetCore> {
  rootSelectors: [string, RootNodeCandidate][]
  resources: ResourceStore<D>
  environment: HydraEnvironment<D>
  fetch: typeof fetch
  Headers: typeof Headers
}

export class Alcaeus<D extends DatasetCore> implements HydraClient<D> {
  public baseUri?: string = undefined

  public rootSelectors: [string, RootNodeCandidate][]

  public defaultHeaders: HeadersInit | ((params: { uri: string }) => HeadersInit | Promise<HeadersInit>) = {}

  public defaultRequestInit = {}

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  public log: (msg: string) => void = () => {
  }

  public cacheStrategy: ResourceCacheStrategy = {...DefaultCacheStrategy}

  public readonly resources: ResourceStore<D>

  private readonly __apiDocumentations: Map<NamedNode, ResourceRepresentation<D, ApiDocumentation<D>>> = $rdf.termMap()

  private readonly environment: HydraEnvironment<D>

  private readonly _fetch: ReturnType<typeof FetchUtil>

  private readonly _headers: typeof Headers

  public constructor(init: AlcaeusInit<D>) {
    const {resources, environment, rootSelectors, fetch} = init

    this.rootSelectors = rootSelectors
    this.resources = resources
    this.environment = environment
    this._headers = init.Headers
    this._fetch = FetchUtil(fetch, this._headers)
  }

  public get apiDocumentations() {
    return [...this.__apiDocumentations.values()]
  }

  public async loadResource<T extends RdfResourceCore<any>>(id: string | NamedNode, headers: HeadersInit = {}, requestInit?: AllowedRequestInit, dereferenceApiDocumentation = true): Promise<HydraResponse<D, T>> {
    const term = typeof id === 'string' ? $rdf.namedNode(id) : id
    let requestHeaders = new this._headers(headers)

    const previousResource = this.resources.get<T>(term)
    if (previousResource) {
      if (!this.cacheStrategy.shouldLoad(previousResource)) {
        return previousResource
      }

      const cacheHeadersInit = this.cacheStrategy.requestCacheHeaders(previousResource)
      if (cacheHeadersInit) {
        const cacheHeaders = new this._headers(cacheHeadersInit)
        requestHeaders = merge(requestHeaders, cacheHeaders, this._headers)
      }
    }

    const uri = getAbsoluteUri(term.value, this.baseUri)
    const defaultRequestInit = typeof this.defaultRequestInit === 'function'
      ? await this.defaultRequestInit({uri})
      : this.defaultRequestInit
    const response = await this._fetch.resource(uri, {
      parsers: this.environment.formats,
      headers: await this.__mergeHeaders(requestHeaders, {uri}),
      ...{...defaultRequestInit, ...requestInit},
    })

    if (previousResource) {
      const etag = response.xhr.headers.get('etag')
      const previousEtag = previousResource.response.xhr.headers.get('etag')
      const etagsEqual = etag && etag === previousEtag
      if (response.xhr.status === 304 || etagsEqual) {
        return previousResource
      }
    }

    const stream = await response.quadStream()
    if (stream) {
      const dataset = await fromStream(this.environment.dataset(), stream)
      const rootResource = this.__findRootResource(dataset, response)

      if (dereferenceApiDocumentation) {
        await this.__getApiDocumentation(response, headers)
      }

      const resources = response.xhr.ok ? this.resources : this.resources.clone()
      await resources.set(term, {
        response,
        dataset,
        rootResource,
      })

      return resources.get<T>(term)!
    }

    return {
      response,
    }
  }

  public async loadDocumentation(id: string | NamedNode, headers: HeadersInit = {}, requestInit: AllowedRequestInit = {}): Promise<ApiDocumentation<D> | null> {
    const term = typeof id === 'string' ? $rdf.namedNode(id) : id

    const resource = await this.loadResource<ApiDocumentation<D>>(term, headers, requestInit, false)
    if (resource && resource.representation) {
      this.__apiDocumentations.set(term, resource.representation)
      const [apiDocs] = resource.representation.ofType<ApiDocumentation<D>>(hydra.ApiDocumentation)
      if (apiDocs) {
        return apiDocs
      }
    }

    return null
  }

  public async invokeOperation<T extends RdfResourceCore<any> = Resource<D>>(operation: InvokedOperation, headers: HeadersInit = {}, body?: BodyInit, requestInit: AllowedRequestInit = {}): Promise<HydraResponse<D, T>> {
    const uri = getAbsoluteUri(operation.target.id.value, this.baseUri)
    const mergedHeaders = await this.__mergeHeaders(new this._headers(headers), {uri})

    if (!operation.method) {
      throw new Error('Cannot invoke operation without a hydra:method')
    }

    const defaultRequestInit = typeof this.defaultRequestInit === 'function'
      ? await this.defaultRequestInit({uri})
      : this.defaultRequestInit
    const response = await this._fetch.operation(operation.method, uri, {
      parsers: this.environment.formats.parsers,
      headers: mergedHeaders,
      body,
      ...{...defaultRequestInit, ...requestInit},
    })
    await this.__getApiDocumentation(response, headers)

    const responseStream = await response.quadStream()

    if (responseStream) {
      const dataset = await fromStream(this.environment.dataset(), responseStream)
      const rootResource = this.__findRootResource(dataset, response)
      let resources = this.resources
      if (operation.method?.toUpperCase() !== 'GET') {
        resources = this.resources.clone()
      }

      await resources.set($rdf.namedNode(response.resourceUri), {
        dataset,
        rootResource,
        response,
      })

      return resources.get<T>($rdf.namedNode(response.resourceUri))!
    }

    return {
      response,
    }
  }

  private async __getApiDocumentation(response: ResponseWrapper, headers: HeadersInit): Promise<void> {
    if (!response.apiDocumentationLink) {
      this.log(`Resource ${response.requestedUri} does not expose API Documentation link`)
      return
    }

    await this.loadDocumentation(response.apiDocumentationLink, headers)
  }

  private async __mergeHeaders(headers: Headers, {uri}: { uri: string }): Promise<Headers> {
    const defaultHeaders = typeof this.defaultHeaders === 'function' ? await this.defaultHeaders({uri}) : this.defaultHeaders

    return merge(new this._headers(defaultHeaders), headers, this._headers)
  }

  private __findRootResource(dataset: D, response: ResponseWrapper): ResourceIdentifier | undefined {
    const candidateNodes = this.rootSelectors.reduce((candidates, [, getCandidate]) => {
      const candidate = getCandidate(response, dataset)
      if (candidate && dataset.match(candidate).size) {
        candidates.add(candidate)
      }

      return candidates
    }, $rdf.termSet<ResourceIdentifier>())

    if (!candidateNodes.size) {
      return undefined
    }

    const candidates = [...candidateNodes].map(term => cf({dataset, term}))

    return candidates.sort(byInProperties)[0].term
  }
}
