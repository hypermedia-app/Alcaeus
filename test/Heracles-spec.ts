import namespace from '@rdfjs/namespace'
import clownface from 'clownface'
import type { DatasetExt } from 'rdf-ext/lib/Dataset.js'
import $rdf from 'rdf-ext'
import { hydra, rdfs } from '@tpluscode/rdf-ns-builders'
import chai, { expect } from 'chai'
import sinon from 'sinon'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { create, PartialCollectionView } from '../src/index.js'
import { HydraClient } from '../src/alcaeus.js'
import FetchUtil from '../src/FetchUtil.js'
import { Bodies } from './test-objects/index.js'
import { mockedResponse, responseBuilder } from './test-utils.js'

// eslint-disable-next-line no-unused-vars
let fetchUtil: ReturnType<typeof FetchUtil>
let invokeOperation: sinon.SinonStub
let fetchResource: sinon.SinonStub

function mockFetchUtil() {
  invokeOperation = sinon.stub()
  fetchResource = sinon.stub()
  fetchUtil = {
    operation: invokeOperation,
    resource: fetchResource,
  }
}

const ex = namespace('http://example.com/')

describe('Hydra loadDocumentation', () => {
  let client: HydraClient
  let dataset: DatasetExt

  chai.use(jestSnapshotPlugin())
  beforeEach(mockFetchUtil)
  beforeEach(() => {
    dataset = $rdf.dataset()
    client = create({
      dataset,
      environment: $rdf,
    })
  })

  it('should store its representation in the dataset', async () => {
    // given
    const body = {
      '@id': 'http://api.example.com/doc/',
      '@type': hydra.ApiDocumentation.value,
    }
    fetchResource.onFirstCall().returns(mockedResponse({
      xhrBuilder: responseBuilder().body(body),
    }))

    // when
    await client.loadDocumentation('http://api.example.com/doc/')

    // then
    expect(dataset.toCanonical()).toMatchSnapshot()
  })

  it('should replace its representation in the dataset when loading twice', async () => {
    // given
    const body = (suffix = '') => ({
      '@id': 'http://api.example.com/doc/' + suffix,
      '@type': hydra.ApiDocumentation.value,
    })
    fetchResource.onFirstCall().returns(mockedResponse({
      xhrBuilder: responseBuilder().body(body('1')),
    }))
    fetchResource.onFirstCall().returns(mockedResponse({
      xhrBuilder: responseBuilder().body(body('2')),
    }))

    // when
    await client.loadDocumentation('http://api.example.com/doc/')
    await client.loadDocumentation('http://api.example.com/doc/')

    // then
    expect(dataset.toCanonical()).toMatchSnapshot()
  })

  it('passes absolute URI to fetch', async () => {
    // given
    fetchResource.onFirstCall().returns(mockedResponse({
      xhrBuilder: responseBuilder().body(''),
    }))
    client.baseUri = 'http://example.com/foo/'

    // when
    await client.loadDocumentation('bar/docs')

    // then
    expect(fetchResource)
      .to.have.been.calledWith(
        'http://example.com/foo/bar/docs',
        sinon.match.any)
  })

  it('forwards additional options to fetch', async () => {
    // given
    fetchResource.onFirstCall().returns(mockedResponse({
      xhrBuilder: responseBuilder().body(''),
    }))
    client.baseUri = 'http://example.com/foo/'

    // when
    await client.loadDocumentation('bar/docs', {}, {
      credentials: 'same-origin',
    })

    // then
    expect(fetchResource)
      .to.have.been.calledWith(
        'http://example.com/foo/bar/docs',
        sinon.match({
          credentials: 'same-origin',
        }))
  })
})

describe('Hydra', () => {
  let loadDocumentation: sinon.SinonStub
  let client: HydraClient
  let dataset: DatasetExt

  beforeEach(mockFetchUtil)
  beforeEach(() => {
    dataset = $rdf.dataset()
    client = create({
      dataset,
      environment: $rdf,
    })
    loadDocumentation = (client.loadDocumentation = sinon.stub().resolves({}))
    fetchResource.reset()
    invokeOperation.reset()
  })

  describe('loadResource', () => {
    it('should return object with matching @id when it is unescaped in response', async () => {
      // given
      const unescaped = 'http://example.com/biała-gęś'
      const id = 'http://example.com/bia%C5%82a-g%C4%99%C5%9B'
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.unescapedDiacritics),
      }))

      // when
      const hydraRes = await client.loadResource(id)
      const res = hydraRes.representation?.get(id)

      // then
      expect(res!.id.value).to.eq(unescaped)
    })

    it('should return object with matching @id when selected with unescaped uri', async () => {
      // given
      const id = 'http://example.com/biała-gęś'
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.unescapedDiacritics),
      }))

      // when
      const hydraRes = await client.loadResource(id)
      const res = hydraRes.representation?.get(id)

      // then
      expect(res!.id.value).to.eq(id)
    })

    it('should load documentation', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      }))

      // when
      await client.loadResource('http://example.com/resource')

      // then
      expect(loadDocumentation).to.have.been.calledWith('http://api.example.com/doc/', {})
    })

    it('should not load documentation in absence of Link header', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        includeDocsLink: false,
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      }))

      // when
      await client.loadResource('http://example.com/resource')

      // then
      expect(fetchResource).to.have.callCount(1)
    })

    it('should load parent of collection view as Resource', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.hydraCollectionWithView),
      }))

      // when
      const hydraRes = await client.loadResource('http://example.com/resource?page=3')
      const res = hydraRes.representation?.get<PartialCollectionView>('http://example.com/resource?page=3')

      // then
      expect(res?.parent).not.to.be.undefined
      expect(res?.parent).not.to.be.null
    })

    it('should load resource with deep blank node structure', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.deepBlankNodes),
      }))

      // when
      const hydraRes = await client.loadResource('http://example.com/root')
      const res = hydraRes.representation?.get('http://example.com/root') as any

      // then
      const p = 'http://example.com/prop'
      const t = 'http://example.com/text'

      expect(res[p][p][p][p][t].value).to.eq('I\'m nested way deep')
    })

    it('should return typed string literals as their values', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.typedLiteral),
      }))

      // when
      const hydraRes = await client.loadResource('http://example.com/resource')
      const res = hydraRes.representation?.get('http://example.com/resource') as any

      // then
      expect(res['http://schema.org/image']['http://schema.org/contentUrl'].value)
        .to.eq('http://wikibus-test.gear.host/book/1936/image')
    })

    it('should return only response when is cannot be parsed', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().notFound(),
      }))

      // when
      const hydraRes = await client.loadResource('http://example.com/resource')

      // then
      expect(hydraRes.representation).to.be.undefined
    })

    it('should not add representation to store if status is not success', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder()
          .notFound()
          .body({
            '@id': 'http://example.com/Foo',
            [rdfs.label.value]: 'Bar',
          }),
      }))

      // when
      await client.loadResource('http://example.com/resource')

      // then
      expect(client.resources.get($rdf.namedNode('http://example.com/resource'))).to.be.undefined
    })

    it.skip('should return typed numeric literals as their values', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.typedNumericLiteral),
      }))

      // when
      const hydraRes = await client.loadResource('http://example.com/resource')
      const res = <any> hydraRes.representation?.get('http://example.com/resource')

      // then
      expect(res['http://schema.org/age']).to.eq(21)
    })

    afterEach(() => {
      loadDocumentation.restore()
      fetchResource.reset()
    })
  })

  describe('default root selectors', () => {
    it('should select by exact id if exists', async () => {
      // given
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      }))

      // when
      const { representation } = await client.loadResource('http://example.com/resource')

      // then
      expect(representation?.root!.id.value).to.eq('http://example.com/resource')
    })

    it.skip('should select resource with redirected id if original is not present', async () => {
      // given
      const redirectUri = 'http://example.com/resource'

      const xhrBuilder = responseBuilder().body(Bodies.someJsonLd).redirect(redirectUri)
      fetchResource.onFirstCall().returns(mockedResponse({ xhrBuilder }))

      // when
      const { representation } = await client.loadResource('http://example.com/not-there')

      // then
      expect(representation?.root!.id).to.eq('http://example.com/resource')
    })
  })

  describe('invokeOperation', () => {
    describe('POST method', () => {
      it('does not store response in dataset', async () => {
        // given
        invokeOperation.onFirstCall().returns(mockedResponse({
          xhrBuilder: responseBuilder().body(Bodies.typedLiteral),
        }))
        const operation = {
          method: 'POST',
          target: {
            id: ex.resource,
          },
        }

        // when
        await client.invokeOperation(operation)

        // then
        expect(dataset).to.have.length(0)
      })

      it('returns data from the response', async () => {
        // given
        invokeOperation.onFirstCall().returns(mockedResponse({
          xhrBuilder: responseBuilder().body(Bodies.typedLiteral),
        }))
        const operation = {
          method: 'POST',
          target: {
            id: ex.resource,
          },
        }

        // when
        const { representation } = await client.invokeOperation(operation)

        // then
        expect(representation?.length).to.be.gt(0)
      })
    })

    describe('GET method', () => {
      it('stores response in dataset', async () => {
        // given
        invokeOperation.onFirstCall().returns(mockedResponse({
          xhrBuilder: responseBuilder().body(Bodies.typedLiteral),
        }))
        const operation = {
          method: 'GET',
          target: {
            id: ex.resource,
          },
        }

        // when
        await client.invokeOperation(operation)

        // then
        expect(dataset.size).to.be.gt(0)
      })
    })
  })

  describe('customizing default request init', () => {
    let client: HydraClient

    beforeEach(() => {
      client = create({ environment: $rdf })
      fetchResource.callsFake(mockedResponse({
        xhrBuilder: responseBuilder().body(''),
      }))
      invokeOperation.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(''),
      }))
    })

    it('combines default init with per-request init', async () => {
      // given
      client.defaultRequestInit = {
        cache: 'no-cache',
        mode: 'cors',
      }

      // when
      await client.loadResource('uri', {}, {
        credentials: 'omit',
        mode: 'no-cors',
      })

      // then
      expect(fetchResource).to.have.been.calledWith(
        'uri',
        sinon.match({
          headers: new Headers({
            Authorization: 'Bearer foobar',
          }),
          cache: 'no-cache',
          credentials: 'omit',
          mode: 'no-cors',
        }))
    })

    it('combines lazy request init defaults with per-request init', async () => {
      // given
      client.defaultRequestInit = async () => ({
        cache: 'no-cache',
        mode: 'cors',
      })

      // when
      await client.loadResource('uri', {}, {
        credentials: 'omit',
        mode: 'no-cors',
      })

      // then
      expect(fetchResource).to.have.been.calledWith(
        'uri',
        sinon.match({
          headers: new Headers({
            Authorization: 'Bearer foobar',
          }),
          cache: 'no-cache',
          credentials: 'omit',
          mode: 'no-cors',
        }))
    })

    it('combines default init with per-request init when invoking operation', async () => {
      // given
      client.defaultRequestInit = {
        cache: 'no-cache',
        mode: 'cors',
      }
      const operation = {
        method: 'POST',
        target: {
          id: ex.resource,
        },
      }

      // when
      await client.invokeOperation(operation, {}, undefined, {
        credentials: 'omit',
        mode: 'no-cors',
      })

      // then
      expect(invokeOperation).to.have.been.calledWith(
        'POST',
        ex.resource.value,
        sinon.match({
          headers: sinon.match({
            Authorization: 'Bearer foobar',
          }),
          cache: 'no-cache',
          credentials: 'omit',
          mode: 'no-cors',
        }))
    })

    it('combines func request init with per-request init when invoking operation', async () => {
      // given
      client.defaultRequestInit = () => ({
        cache: 'no-cache',
        mode: 'cors',
      })
      const operation = {
        method: 'POST',
        target: {
          id: ex.resource,
        },
      }

      // when
      await client.invokeOperation(operation, {}, undefined, {
        credentials: 'omit',
        mode: 'no-cors',
      })

      // then
      expect(invokeOperation).to.have.been.calledWith(
        'POST',
        ex.resource.value,
        sinon.match({
          headers: sinon.match({
            Authorization: 'Bearer foobar',
          }),
          cache: 'no-cache',
          credentials: 'omit',
          mode: 'no-cors',
        }))
    })
  })

  describe('customizing default headers', () => {
    let client: HydraClient

    beforeEach(() => {
      client = create({ environment: $rdf })
      fetchResource.callsFake(mockedResponse({
        xhrBuilder: responseBuilder().body(''),
      }))
      invokeOperation.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().body(''),
      }))
    })

    describe('as HeadersInit object', () => {
      it('passes them to loadResource', async () => {
        // given
        client.defaultHeaders = {
          Authorization: 'Bearer foobar',
        }

        // when
        await client.loadResource('uri')

        // then
        expect(fetchResource).to.have.been.calledWith(
          'uri',
          sinon.match({
            headers: new Headers({
              Authorization: 'Bearer foobar',
            }),
          }))
      })

      it('passes them to invokeOperation', async () => {
        // given
        client.defaultHeaders = {
          Authorization: 'Bearer foobar',
        }
        const operation = {
          method: 'post',
          target: {
            id: ex.uri,
          },
        }

        // when
        await client.invokeOperation(operation)

        // then
        expect(invokeOperation)
          .to.have.been.calledWith(
            'post',
            ex.uri.value,
            sinon.match({
              headers: new Headers({
                Authorization: 'Bearer foobar',
              }),
            }))
      })

      it('passes them to loadDocumentation', async () => {
        // given
        client.defaultHeaders = {
          Authorization: 'Bearer foobar',
        }

        // when
        await client.loadDocumentation('doc')

        // then
        expect(fetchResource).to.have.been.calledWith(
          'doc', sinon.match({
            headers: new Headers({
              Authorization: 'Bearer foobar',
            }),
          }))
      })
    })

    describe('as HeadersInit function', () => {
      it('passes them to loadResource', async () => {
        // given
        client.defaultHeaders = () => ({
          Authorization: 'Token xyz',
        })

        // when
        await client.loadResource('uri')

        // then
        expect(fetchResource).to.have.been.calledWith(
          'uri', sinon.match({
            headers: new Headers({
              Authorization: 'Token xyz',
            }),
          }))
      })

      it('passes them to loadDocumentation', async () => {
        // given
        client.defaultHeaders = () => ({
          Authorization: 'Token xyz',
        })

        // when
        await client.loadDocumentation('doc')

        // then
        expect(fetchResource).to.have.been.calledWith(
          'doc', sinon.match({
            headers: new Headers({
              Authorization: 'Token xyz',
            }),
          }))
      })

      it('passes them to invokeOperation', async () => {
        // given
        client.defaultHeaders = () => ({
          Authorization: 'Token xyz',
        })
        const operation = {
          method: 'post',
          target: {
            id: ex.uri,
          },
        }

        // when
        await client.invokeOperation(operation)

        // then
        expect(invokeOperation)
          .to.have.been.calledWith(
            'post',
            ex.uri.value,
            sinon.match({
              headers: new Headers({
                Authorization: 'Token xyz',
              }),
            }))
      })
    })

    describe('as async HeadersInit function', () => {
      it('passes them to loadResource', async () => {
        // given
        client.defaultHeaders = async () => ({
          Authorization: 'Token xyz',
        })

        // when
        await client.loadResource('uri')

        // then
        expect(fetchResource).to.have.been.calledWith(
          'uri', sinon.match({
            headers: new Headers({
              Authorization: 'Token xyz',
            }),
          }))
      })

      it('passes them to loadDocumentation', async () => {
        // given
        client.defaultHeaders = async () => ({
          Authorization: 'Token xyz',
        })

        // when
        await client.loadDocumentation('doc')

        // then
        expect(fetchResource).to.have.been.calledWith(
          'doc', sinon.match({
            headers: new Headers({
              Authorization: 'Token xyz',
            }),
          }))
      })

      it('passes them to invokeOperation', async () => {
        // given
        client.defaultHeaders = async () => ({
          Authorization: 'Token xyz',
        })
        const operation = {
          method: 'post',
          target: {
            id: ex.uri,
          },
        }

        // when
        await client.invokeOperation(operation)

        // then
        expect(invokeOperation)
          .to.have.been.calledWith(
            'post',
            ex.uri.value,
            sinon.match({
              headers: new Headers({
                Authorization: 'Token xyz',
              }),
            }))
      })

      it('receives requested uri', async () => {
        // given
        client.defaultHeaders = sinon.stub()

        // when
        await client.loadResource('http://example.com/resource')

        // then
        expect(client.defaultHeaders).to.have.been.calledWith({
          uri: 'http://example.com/resource',
        })
      })

      it('receives absolute uri when using baseUri setting', async () => {
        // given
        client.defaultHeaders = sinon.stub()
        client.baseUri = 'http://example.com/'

        // when
        await client.loadResource('resource')

        // then
        expect(client.defaultHeaders).to.have.been.calledWith({
          uri: 'http://example.com/resource',
        })
      })

      it('receives absolute uri when using baseUri setting from operation', async () => {
        // given
        client.defaultHeaders = sinon.stub()
        client.baseUri = 'http://example.com/'
        const operation = {
          method: 'post',
          target: {
            id: $rdf.namedNode('resource'),
          },
        }

        // when
        await client.invokeOperation(operation)

        // then
        expect(client.defaultHeaders).to.have.been.calledWith({
          uri: 'http://example.com/resource',
        })
      })
    })
  })

  describe('loadResource + cache strategy', () => {
    const id = 'http://example.com/resource'

    it('should not do a request when resource is used from store', async () => {
      // given
      const responseMock = mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      })
      fetchResource.onFirstCall().returns(responseMock)
      client.cacheStrategy.shouldLoad = () => false
      clownface({ dataset }).namedNode(id).addOut(rdfs.label, 'Foo')
      await client.resources.set($rdf.namedNode(id), {
        response: await responseMock(id),
        dataset,
      })

      // when
      await client.loadResource(id)

      // then
      expect(fetchResource).not.to.to.have.been.called
    })

    it('should reuse previous representation when server responds with 304', async () => {
      // given
      const previousResponse = await mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      })(id)
      fetchResource.onFirstCall().returns(mockedResponse({
        xhrBuilder: responseBuilder().statusCode(304),
      }))
      client.cacheStrategy.shouldLoad = () => true
      clownface({ dataset }).namedNode(id).addOut(rdfs.label, 'Foo')
      await client.resources.set($rdf.namedNode(id), {
        response: previousResponse,
        dataset,
      })

      // when
      const res = await client.loadResource(id)

      // then
      expect(res.representation).to.be.ok
      expect(res.response).to.eq(previousResponse)
    })

    it('should reuse previous representation when server responds with equal etag', async () => {
      // given
      const previousResponse = await mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd).header('etag', 'foobar'),
      })(id)
      fetchResource.callsFake(mockedResponse({
        xhrBuilder: responseBuilder().statusCode(200).header('etag', 'foobar'),
      }))
      client.cacheStrategy.shouldLoad = () => true
      clownface({ dataset }).namedNode(id).addOut(rdfs.label, 'Foo')
      await client.resources.set($rdf.namedNode(id), {
        response: previousResponse,
        dataset,
      })

      // when
      const res = await client.loadResource(id)

      // then
      expect(res.representation).not.to.be.undefined
      expect(res.response).to.eq(previousResponse)
    })

    it('should set caching request headers provided by cache strategy', async () => {
      // given
      const previousResponse = mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      })
      fetchResource.onFirstCall().returns(previousResponse)
      client.cacheStrategy.shouldLoad = () => true
      client.cacheStrategy.requestCacheHeaders = () => ({
        'if-none-match': 'foo',
      })
      clownface({ dataset }).namedNode(id).addOut(rdfs.label, 'Foo')
      await client.resources.set($rdf.namedNode(id), {
        response: await previousResponse(id),
        dataset,
      })

      // when
      await client.loadResource(id)

      // then
      expect(fetchResource).to.have.been.calledWith(
        id,
        sinon.match({
          headers: new Headers({ 'if-none-match': 'foo' }),
        }))
    })

    it('should replace cached resource when cache strategy returns true', async () => {
      // given
      const responseMock = mockedResponse({
        xhrBuilder: responseBuilder().body(Bodies.someJsonLd),
      })
      fetchResource.onFirstCall().returns(responseMock)
      client.cacheStrategy.shouldLoad = () => true
      const term = $rdf.namedNode(id)
      clownface({ dataset, graph: term, term }).addOut(rdfs.label, 'Foo')
      await client.resources.set($rdf.namedNode(id), {
        response: await responseMock(id),
        dataset,
      })

      // when
      await client.loadResource(id)

      // then
      expect(fetchResource).to.to.have.been.called
      expect(dataset.toCanonical()).toMatchSnapshot()
    })
  })
})
