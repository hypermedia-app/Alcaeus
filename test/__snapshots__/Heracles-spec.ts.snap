// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hydra loadDocumentation should replace its representation in the dataset when loading twice 1`] = `
"<http://api.example.com/doc/2> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#ApiDocumentation> <http://api.example.com/doc/> .
"
`;

exports[`Hydra loadDocumentation should store its representation in the dataset 1`] = `
"<http://api.example.com/doc/> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#ApiDocumentation> <http://api.example.com/doc/> .
"
`;

exports[`Hydra loadResource + cache strategy should replace cached resource when cache strategy returns true 1`] = `
"<http://example.com/resource> <http://example.com/vocab#other> <http://example.com/linked> <http://example.com/resource> .
<http://example.com/resource> <http://example.com/vocab#other_yet> <http://example.com/linked> <http://example.com/resource> .
<http://example.com/resource> <http://example.com/vocab#prop> \\"some textual value\\" <http://example.com/resource> .
<http://example.com/resource> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/vocab#Resource> <http://example.com/resource> .
"
`;
